src/pdal/PyArray.cpp

```diff
@@ -164,51 +164,85 @@ Array::~Array()
     Py_XDECREF(m_array);
 }
 
-
-ArrayIter& Array::iterator()
+std::shared_ptr<ArrayIter> Array::iterator()
 {
-    ArrayIter *it = new ArrayIter(m_array);
-    m_iterators.emplace_back((it));
-    return *it;
+    return std::make_shared<ArrayIter>(m_array, m_stream_handler);
 }
 
+ArrayIter::ArrayIter(PyArrayObject* np_array, std::shared_ptr<ArrayStreamHandler> stream_handler)
+    : m_stream_handler(std::move(stream_handler))
+{
+    resetIterator(np_array);
+}
 
-ArrayIter::ArrayIter(PyArrayObject* np_array)
+void ArrayIter::resetIterator(std::optional<PyArrayObject*> np_array = {})
```

Comment:
It seems like there are really two separate cases here. An `initialize` case and a `reset` case, variously triggered by the optional argument. Perhaps more clear to have two functions?

Reply:
Hmm. This didn't come out as nicely as I was thinking, so I may regret my comment ;)